cdap-notifications/src/main/java/co/cask/cdap/notifications/service/inmemory/InMemoryNotificationService.java

```diff
@@ -0,0 +1,180 @@
+import com.google.common.util.concurrent.MoreExecutors;
+import com.google.inject.Inject;
+import org.apache.twill.common.Cancellable;
+import org.apache.twill.common.Threads;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.lang.reflect.Type;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.Executor;
+import java.util.concurrent.Executors;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+
+/**
+ * In memory Notification service that pushes notifications to subscribers.
+ */
+public class InMemoryNotificationService extends AbstractIdleService implements NotificationService {
+  private static final Logger LOG = LoggerFactory.getLogger(InMemoryNotificationService.class);
+
+  private final SetMultimap<NotificationFeed, NotificationHandler> feedsToHandlers;
+  private final Map<NotificationHandler, Executor> handlersToExecutors;
+  private final ReadWriteLock handlersLock;
+
+  private final DatasetFramework dsFramework;
+  private final TransactionSystemClient transactionSystemClient;
+  private final NotificationFeedManager feedManager;
+
+  private ListeningExecutorService publishingExecutor;
+
+  @Inject
+  public InMemoryNotificationService(DatasetFramework dsFramework, TransactionSystemClient transactionSystemClient,
+                                     NotificationFeedManager feedManager) {
+    this.dsFramework = dsFramework;
+    this.transactionSystemClient = transactionSystemClient;
+    this.feedManager = feedManager;
+    this.feedsToHandlers = HashMultimap.create();
+    this.handlersToExecutors = Maps.newHashMap();
+    this.handlersLock = new ReentrantReadWriteLock();
+  }
+
+  @Override
+  protected void startUp() throws Exception {
+    publishingExecutor = MoreExecutors.listeningDecorator(
+      Executors.newSingleThreadExecutor(Threads.createDaemonThreadFactory("notification-publisher")));
+  }
+
+  @Override
+  protected void shutDown() throws Exception {
+    publishingExecutor.shutdownNow();
+  }
+
+  @Override
+  public <N> ListenableFuture<N> publish(NotificationFeed feed, N notification)
+    throws NotificationException, NotificationFeedException {
+    return publish(feed, notification, notification.getClass());
+  }
+
+  @Override
+  public <N> ListenableFuture<N> publish(final NotificationFeed feed, final N notification, Type notificationType)
+    throws NotificationException, NotificationFeedException {
+    handlersLock.readLock().lock();
+    try {
+      LOG.debug("Publishing on notification feed [{}]: {}", feed, notification);
+      Set<NotificationHandler> handlers = feedsToHandlers.get(feed);
+      if (handlers.isEmpty()) {
+        return Futures.immediateFuture(null);
+      }
+
+      for (final NotificationHandler handler : handlers) {
+        // We use an executor here so that we can release the lock before calling the users' executors
```

Comment:
Better to have a class to encapsulate the calling logic. Also, you can simplify the logic without using explicit lock. Something like this:

```java
Multimap<NotificationFeed, NotificationCaller<?>> subscribers = Multimaps.synchronizedMultimap(HashMultimap.create());
...

public <N> Cancellable subscribe(NotificationFeed feed, NotificationHandler<N> handler, Executor executor) {
  final NotificationCaller<N> caller = new NotificationCaller(handler, executor);
  subscribers.put(feed, caller);
  return caller;
}

public <N> ListenableFuture<N> publish(NotificationFeed feed, N notification, Type notificationType) {
  List<NotificationCaller> callers = ImmutableList.copyOf(subscribers.get(feed));
  for (NotificationCaller caller : callers) {
    caller.received(notification, new BasicNotificationContext(...));
  }
}

class NotificationCaller<N> implements NotificationHandler<N>, Cancellable {
  private final NotificationFeed feed;
  private final NotificationHandler<N> handler;
  private final Executor executor;
  private volatile boolean completed;

  NotificationCaller(NotificationFeed feed, NotificationHandler<N> handler, Executor executor) {
    this.feed = feed;
    this.handler = handler;
    this.executor = executor;
  }
  
  public void received(N notification, NotificationContext notificationContext) {
    if (completed) {
      return;
    }
    executor.execute(new Runnable() {
      public void run() {
        if (completed) {
          return;
        }
        try {
          handler.received(notification, notificationContext);
        } catch (Throwable t) {
          ....
        }
      }
    });
  }

  public void cancel() {
    completed = true;
    subscibers.remove(feed, this);
  }
}
```